pca_cca.go

```diff
@@ -0,0 +1,262 @@
+// and y (column means subtracted from each column), let Sx and Sy denote the
+// sample covariance matrices within x and y respectively, and let Sxy denote
+// the covariance matrix between x and y. The sphered data can then be expressed
+// as Xc * Sx^(-1/2) and Yc * Sy^(-1/2) respectively, and the correlation matrix
+// between the sphered data is called the canonical correlation matrix,
+// Sx^(-1/2) * Sxy * Sy^(-1/2). In cases where S^(-1/2) is ambiguous for some
+// covariance matrix S, S^(-1/2) is taken to be E * D^(-1/2) * E' where S can be
+// eigendecomposed as S = E * D * E'.
+//
+// The canonical correlations are the correlations between the corresponding
+// pairs of canonical variables and can be obtained with c.Corrs(). Canonical
+// variables can be obtained by projecting the sphered data into the left and
+// right eigenvectors of the canonical correlation matrix, and these
+// eigenvectors can be obtained with c.Left(true) and c.Right(true)
+// respectively. The canonical variables can also be obtained directly from the
+// centered raw data by using the back-transformed eigenvectors which can be
+// obtained with c.Left(false) and c.Right(false) respectively.
+//
+// The first pair of left and right eigenvectors of the canonical correlation
+// matrix can be interpreted as directions into which the respective sphered
+// data can be projected such that the correlation between the two projections
+// is maximised. The second pair and onwards solve the same optimization but
+// under the constraint that they are uncorrelated (orthogonal in sphered space)
+// to previous projections.
+//
+// CanonicalCorrelations will panic if the inputs x and y do not have the same
+// number of rows. Columns of x and y should be interpretable as two different
+// sets of measurements on the same observations.
+//
+// The slice weights is used to weight the observations. If weights is nil, each
+// weight is considered to have a value of one, otherwise the length of weights
+// must match the number of observations or CanonicalCorrelations will panic.
+//
+// More details can be found at
+// https://en.wikipedia.org/wiki/Canonical_correlation
+// or in Chapter 3 of
+// Koch, Inge. Analysis of multivariate and high-dimensional data.
+// Vol. 32. Cambridge University Press, 2013. ISBN: 9780521887939
+func CanonicalCorrelations(x, y mat64.Matrix, weights []float64) (c CC, err error) {
+	n, _ := x.Dims()
+	yn, _ := y.Dims()
+	if n != yn {
+		panic("stat: unequal number of observations")
+	}
+	if weights != nil && len(weights) != n {
+		panic("stat: len(weights) != observations")
+	}
+
+	// Center and factorize x and y.
+	xsvd, ok := svdFactorizeCentered(x, weights)
+	if !ok {
+		return CC{}, errors.New("stat: failed to factorize x")
+	}
+	ysvd, ok := svdFactorizeCentered(y, weights)
+	if !ok {
+		return CC{}, errors.New("stat: failed to factorize y")
+	}
+	var xu, xv, yu, yv mat64.Dense
+	xu.UFromSVD(xsvd)
```

Comment:
Good point!

This also reminds me of something I noticed in PrincipalComponents but forgot to fix -- `n < d`
should actually be `n - 1 < d` and everwhere that `min(n,d)` is used should be `min(n-1,d)` because the columns are centred this affects the upper bound on the rank of the matrix. I'll make this fix. 

While we're talking about rank -- although this is a correct upper bound it is not neccessarily correct, and I believe the usual way of estimating the rank of a matrix is using an epsilon around zero to could how many non-zero singular values it has. Is this something worth considering/ that would be useful?